src/pybind/cephfs/cephfs.pyx

```diff
@@ -2832,3 +2846,288 @@ cdef class LibCephFS(object):
+        If it fails, add this new dir to current dir's ignorelist
+        '''
+        de_path = os.path.join(self.curr_dir.path, de_name)
+        try:
+            self.stack.append(RmtreeDir(self.fs, de_path))
+        # RmtreeDir() eventually/indirectly calls opendir(). If it fails, it is
+        # most likely due to lack of permissions on de_path. Skip it and proceed
+        # deleting rest of files in current directories.
+        except Error as e:
+            log.error(f'Exception occured: "{e}"')
+            self.curr_dir.add_to_de_ignore_list(de_name)
+
+            if self.suppress_errors:
+                return 'break'
+            else:
+                raise
+
+        # New dir has been added to the stack, call rewinddir() since the new
+        # dir will be traversed now. Else, future call to current dir's
+        # readdir() will have implementation-defined behaviour, which might be
+        # different from the expected behaviour.
+        self.curr_dir.handle.rewinddir()
+
+        # since a new non-empty dir has been encountered abort traversing
+        # current dir for now and traverse this new dir instead
+        return 'break'
+
+    def notify_parent_dir(self):
+        '''
+        Add current dir's name to parent dir's "de_ignore_list". This is \
+        necessary since parent dir can't be deleted when current dir can't be
+        deleted.
+        '''
+        # ensure we are dealing with the dir at the top of the stack.
+        assert self.curr_dir is self.stack[-1]
+
+        try:
+            parent_dir = self.stack[-2]
+            # Save as byte instead of str since it is de.d_name fetched by
+            # readdir() is always bytes.
+            parent_dir.add_to_de_ignore_list(self.curr_dir.name)
+        except IndexError:
+            pass
+
+    def rmtree(self):
+        '''
+        This is where depth-first, non-recursive traversal is done.
+        '''
+        try:
+            self.stack.append(RmtreeDir(self.fs, self.trash_path))
+        except Exception as e:
+            log.error('opening root dir of the file tree failed with exception '
+                      f'"{e}", exiting.')
+            if self.suppress_errors:
+                return
+            else:
+                raise
+
+        while self.stack:
+            if self.should_cancel():
+                raise OpCancelled('rmtree')
+
+            self.curr_dir = self.stack[-1]
+
+            # de = directory entry
+            de = self.curr_dir.read_dir()
+            while de:
+                if self.should_cancel():
+                    raise OpCancelled('rmtree')
+
+                de_name = de.d_name.decode()
+                if de.is_dir():
+                    try:
+                        retval = self.curr_dir.try_rmdir(de_name,
+                                                         self.suppress_errors)
+                    except ObjectNotEmpty:
+                        retval = self.add_dir_to_stack(de_name)
```

Comment:
If this returns `"break"` then it will stop processing the directory, right? That should not be correct.

I think you can structure this so that this `break` return is not necessary. You already removed it for `try_unlink`.

Reply:
> If this returns "break" then it will stop processing the directory, right? That should not be correct.

Done.

> I think you can structure this so that this break return is not necessary. You already removed it for try_unlink.

We are not returning break anymore but we'll need to indicate whether adding a new dir to stack failed or succeeded, based on which the iteration in the inner loop can be halted or allowed to continue. We are doing that now by returning boolean value which is an improvement in comparison.